vulcan/builder.py

```diff
@@ -1,40 +1,57 @@
+                       ) -> Tuple[List[str], Dict[str, List[str]]]:
 
     if not install_requires and not extras:
         return [], {}
 
     extras_list = list(extras.items())
     with create_venv(python_version) as pipenv:
-        with tempfile.TemporaryDirectory() as site_packages:
-            print("Building default requirements")
-            pipenv.install(site_packages, install_requires)
-            base_freeze = pipenv.freeze(site_packages)
-            if not extras_list:
-                # if we have no extras, we are done here.
-                return sorted([str(req) for req in base_freeze.values()]), {}
-
-        with tempfile.TemporaryDirectory() as site_packages:
-            print("Building final resolved packages")
-            pipenv.install(site_packages,
-                           install_requires + list(chain.from_iterable(reqs for _, reqs in extras_list)))
-            all_resolved = pipenv.freeze(site_packages)
-
+        print("Building base requires")
+        base_freeze = await build_requires(pipenv, install_requires)
+        if not extras_list:
+            # if we have no extras, we are done here.
+            return sorted([str(req) for req in base_freeze.values()]), {}
+
+        print("Building requirements for base + all extras")
+        final_out_task = asyncio.get_event_loop().create_task(
+            build_requires(
+                pipenv,
+                install_requires + list(chain.from_iterable(reqs for _, reqs in extras_list))))
         resolved_extras = {}
         for extra, extra_reqs in extras_list:
-            with tempfile.TemporaryDirectory() as site_packages:
-                # this is the expensive bit, because we create a new venv for each extra
-                print(f"Building requirements for extra '{extra}'")
-                pipenv.install(site_packages, install_requires + extra_reqs)
-                extra_freeze = pipenv.freeze(site_packages)
-                resolved_extras[extra] = sorted([str(all_resolved[req]) for req in extra_freeze])
-
-        return sorted([str(all_resolved[req]) for req in base_freeze.keys()]), resolved_extras
+            # this is the expensive bit, because we create a new venv for each extra
+            print(f"Building requirements for extra '{extra}'")
+            resolved_extras[extra] = asyncio.get_event_loop().create_task(
+                build_requires(pipenv, install_requires + extra_reqs))
+
+        # It is important here to wait until ALL tasks are complete (return_exceptions=True) because on
+        # windows, if that is not done then the TemporaryDirectory is create_venv will try to remove itself
+        # while our async subprocesses still have a lock on the python.exe (windows-only issue). which then
+        # causes more errors
+        results = await asyncio.gather(*resolved_extras.values(), final_out_task, return_exceptions=True)
+        for res in results:
+            if isinstance(res, BaseException):
```

Comment:
wouldn't you want to tell users about all the errors? throwing the first only is not so nice I think

Reply:
I agree it's not nice, but good, useful multiexceptions are really difficult. I could maybe aggregate the error messages, but in doing so I'd lose the stack trace completely.

Reply:
positively surprised this works the same way for the asyncio.gather :)